pageserver/src/layered_repository.rs

```diff
@@ -1633,12 +1629,13 @@ impl LayeredTimeline {
     pub fn gc_timeline(&self, retain_lsns: Vec<Lsn>, cutoff: Lsn) -> Result<GcResult> {
         let now = Instant::now();
         let mut result: GcResult = Default::default();
+        let _checkpoint_cs = self.checkpoint_cs.lock().unwrap();
```

Comment:
Do we need to hold `checkpoint_cs`, when we now have the other Mutex? I guess we do, because checkpoint doesn't hold the other Mutex. This is pretty confusing. Can we uphold the invariant that all callers of get_page_at_lsn() and the other Repository get-functions must hold the `latest_gc_cutoff_lsn` lock, and maybe even add an assertion for that?

Reply:
Originally `checkpoint_cs` was introduced to prevent concurrent checkpoints.
As far as neither `gc_cs`, neither  `latest_gc_cutoff_lsn` can prevent it, we still need this mutex.

Is it actually needed to lock in `gc_timeline`? Can we just hold `latest_gc_cutoff_lsn` instead? I am not sure.
`latest_gc_cutoff_lsn`  is not actually blocking GC. It just  blocks start of GC. But if it is already started and advance `latest_gc_cutoff_lsn`  then it still can interfer with GC.
I prefer to hold this mutex here. And I do not think that mutual exclusion of GC and checkpointer will have critical impact on performance.